inference-engine/src/multi_device/multi_device_plugin.cpp

```diff
@@ -403,95 +401,104 @@ DeviceInformation MultiDeviceInferencePlugin::SelectDevice(const std::vector<Dev
+            if (kvp.first >= priority) {
+                continue;
             }
+            auto& filterDevices = kvp.second;
+            auto sd = std::remove_if(validDevices.begin(), validDevices.end(), [&filterDevices](DeviceInformation device) {
+                    auto iter = std::find_if(filterDevices.begin(), filterDevices.end(), [&device](std::string fullName) {
+                            return (fullName == device.deviceName) || (fullName == device.fullName);
+                            });
+                    return iter != filterDevices.end() ? true : false;
+                    });
+            validDevices.erase(sd, validDevices.end());
         }
     }
 
-    // If network is FP32 but there is no device support FP32, offload FP32 network to device support FP16.
-    if (networkPrecision == "FP32") {
-        if (!dGPU.empty()) {
-            for (auto&& item : dGPU) {
-                std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-                auto supportNetwork = std::find(capability.begin(), capability.end(), "FP16");
-                if (supportNetwork != capability.end()) {
-                    return item;
-                }
-            }
-        } else if (!VPUX.empty()) {
-            for (auto&& item : VPUX) {
-                std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-                auto supportNetwork = std::find(capability.begin(), capability.end(), "FP16");
-                if (supportNetwork != capability.end()) {
-                    return item;
-                }
-            }
-        } else if (!iGPU.empty()) {
-            for (auto&& item : iGPU) {
-                std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-                auto supportNetwork = std::find(capability.begin(), capability.end(), "FP16");
-                if (supportNetwork != capability.end()) {
-                    return item;
-                }
-            }
-        } else if (!MYRIAD.empty()) {
-            for (auto&& item : MYRIAD) {
-                std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
-                auto supportNetwork = std::find(capability.begin(), capability.end(), "FP16");
-                if (supportNetwork != capability.end()) {
-                    return item;
-                }
-            }
+    DeviceInformation* ptrSelectDevice =  NULL;
+    if (validDevices.empty()) {
+        // after remove higher priority device,but the available devices is null,
+        // so select the last device of all available Devices.
+        ptrSelectDevice = &lastDevice;
+    } else {
+        // select the first device in the rest of available devices.
+        ptrSelectDevice = &validDevices.front();
```

Comment:
Could we also check if there is free device which is same type as first device?

For example, there is two dGPU, dGPU0 and dGPU1. Network0 select dGPU0 since dGPU0 is first device of list. Network1 is same priority of Network0. We expect Network1 select dGPU1 since dGPU1 is free.

Reply:
I think that the user should be able to specify fully0qualified priorities as well (with device ids), e.g.  explicit "GPU.0" or "GPU.1"

but in case it has specified just "GPU", I agree we need some logic (as e.g. GPU.0 is usually iGPU, then, if presented, the GPU.1 is likely stronger and we shall select that)/
luckily it is possible to deduce the integrated vs descrete info (as the GPU supports the DEVICE_TYPE metric for the specific device name like  "GPU.0"

Reply:
Thanks Maxim. Yes, current device list order is (d)GPU2, (d)GPU1, (i)GPU0. I updated example in previous comment for more accurate.

This example exist on VPU case. If we cannot implement in this PR. We need another PR to follow up it.